crates/iceberg/src/spec/values.rs

```diff
@@ -2334,9 +2453,20 @@ mod _serde {
                 },
                 RawLiteralEnum::Float(v) => match ty {
                     Type::Primitive(PrimitiveType::Float) => Ok(Some(Literal::float(v))),
+                    Type::Primitive(PrimitiveType::Double) => {
+                        Ok(Some(Literal::double(f64::from(v))))
+                    }
                     _ => Err(invalid_err("float")),
                 },
                 RawLiteralEnum::Double(v) => match ty {
+                    Type::Primitive(PrimitiveType::Float) => {
+                        let v = v as f32;
+                        // Return error if the value has lost precision
+                        if v as f64 != f64::from(v) {
```

Comment:
I think this would fail for `NaN` as `NaN` != `NaN`

Reply:
Surely in almost all cases precision will be lost unless the last 29 bits of the `f64`'s fraction are all zeroes (which is about 1 in half a billion), as this comparison will be bitwise?

Should we have special checks for NaN / -Inf / +Inf (possibly -0.0 also?) and for other numbers check that the value is within f32::EPSILON of the original value?

Reply:
> Surely in almost all cases precision will be lost unless the last 29 bits of the f64's fraction are all zeroes (which is about 1 in half a billion), as this comparison will be bitwise?

I think the check here is to deal with the case that serialize f32 to f64 and deserialize back to f32. In this case, there is no precision loss.
For the case loss precision, which means that the deserializer want to deserialize a f64 to f32. I think this means something wrong happen. 🤔

> Should we have special checks for NaN / -Inf / +Inf (possibly -0.0 also?) 

I agree with this.

Reply:
> These are special values, see https://doc.rust-lang.org/std/primitive.f64.html#associatedconstant.NAN we can check and convert them directly: https://doc.rust-lang.org/std/primitive.f32.html#associatedconstant.NAN

If a f64 can convert to f32 losslessly, for `(v1)f64 -> (v2)f32 -> (v3)f64`, bit pattern of v1 and v3 will be same. I think this case can be extended to all values, no matter whether it's Nan, -0, or normal value, so that we can avoid branch here.🤔

Reply:
> TBH I'm not that 100% sure for this, especially on different platforms things may become more akward.

Oh I see. How about the rule like this:
```
let v = v as f32;

if v == NaN || v == -Inf || v == Inf || v == -0.0 || v == 0.0:
  return v;
else:
  let nv:f64 = v.into();
  if |nv - v| >   f32::EPSILON:
    // there is precisison loss
    return err;
```

Reply:
> I think the check here is to deal with the case that serialize f32 to f64 and deserialize back to f32. In this case, there is no precision loss.
For the case loss precision, which means that the deserializer wants to deserialize a f64 to f32. I think this means something wrong happened. 🤔
I think here, what we need to check is whether this f64 is compatible with f32, or says, whether this f64 can be converted to f32 without losing precision. This means that this f64 can convert from a f32.

Thanks for the clarification - that makes total sense 😄 

> So I think we need to do the bitwise comparison here rather than compare two f64. We can convert them to u64 to achieve this.

Not quite, as there are many different bit patterns that correspond to NaN, so I think that one needs to be handled separately.

> Oh I see. How about the rule like this:

> let v = v as f32;
> 
> if v == NaN || v == -Inf || v == Inf || v == -0.0 || v == 0.0:
>   return v;
> else:
>   let nv:f64 = v.into();
>   if |nv - v| >   f32::EPSILON:
>     // there is precisison loss
>     return err;

Works for me (with the caveat that you'll need to use `is_nan` rather than ` == NAN` for it to work)